src/core/license.rs

```diff
@@ -401,66 +426,94 @@ impl BinaryBlob {
+type ServerUpgradeLicense = ServerNewLicense;
 
 /// License data that has been obtained from the sever
 #[allow(dead_code)]
 struct License {
-    data: Vec<u8>,
+    // The version, scope, company name, and product ID fields are used
+    // by the client to index the licenses in the client's license store.
+    version_major: u16,
+    version_minor: u16,
+    scope: String,
+    company_name: String,
+    product_id: String,
+
+    // In theory, this is a DER-encoded X.509 certificate that has some
+    // helpful data in its certificate extensions:
+    //
+    // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpele/99e6bdc5-4517-45a3-be72-a254e8389546
+    //
+    // In practice, something about this cert is non-standard: both Go's crypto/x509 package
+    // and Rust's x509_parser crate choke when attempting to parse it.
+    // (Note that `openssl asn1parse -inform DER` does work)
+    cert_data: Vec<u8>,
 }
 
 impl License {
     fn new(
         session_encryption_data: &SessionEncryptionData,
-        new_license: &NewLicense,
+        new_license: &ServerNewLicense,
     ) -> RdpResult<Self> {
-        let mut rc4 = Rc4::new(Key::<rc4::consts::U16>::from_slice(
-            &session_encryption_data.license_encryption_key,
-        ));
-
-        let mut data: Vec<u8> = new_license.encrypted_license_data.clone();
-        rc4.apply_keystream(&mut data);
-
+        // Decrypt and validate the MAC
+        let data = session_encryption_data.decrypt_message(&new_license.encrypted_license_data);
         if session_encryption_data.generate_mac_data(&data) != new_license.mac_data {
             return Err(Error::RdpError(RdpError::new(
                 RdpErrorKind::InvalidData,
-                "license MAC is different than MAC from encrypted message",
+                "new license MAC is different than MAC from encrypted message",
             )));
         }
-        Ok(Self { data })
+
+        Self::from_decrypted_data(&mut Cursor::new(&data))
+    }
+
+    fn from_decrypted_data(reader: &mut dyn Read) -> RdpResult<Self> {
+        // The decrypted data contains a New License Information packet.
+        // (See MS-RDPELE section 2.2.2.6.1)
+        let mut license_info = component![
+            "dwVersion" => U32::LE(0),
+            "cbScope" => DynOption::new(U32::LE(0), | size | MessageOption::Size("pbScope".to_string(), size.inner() as usize)),
+            "pbScope" => Vec::<u8>::new(),
+            "cbCompanyName" => DynOption::new(U32::LE(0), | size | MessageOption::Size("pbCompanyName".to_string(), size.inner() as usize)),
+            "pbCompanyName" => Vec::<u8>::new(),
+            "cbProductId" => DynOption::new(U32::LE(0), | size | MessageOption::Size("pbProductId".to_string(), size.inner() as usize)),
+            "pbProductId" => Vec::<u8>::new(),
+            "cbLicenseInfo" => DynOption::new(U32::LE(0), | size | MessageOption::Size("pbLicenseInfo".to_string(), size.inner() as usize)),
```

Comment:
`rustfmt` can't format through macros :(
```suggestion
            "cbScope" => DynOption::new(U32::LE(0), |size| MessageOption::Size("pbScope".to_string(), size.inner() as usize)),
            "pbScope" => Vec::<u8>::new(),
            "cbCompanyName" => DynOption::new(U32::LE(0), |size| MessageOption::Size("pbCompanyName".to_string(), size.inner() as usize)),
            "pbCompanyName" => Vec::<u8>::new(),
            "cbProductId" => DynOption::new(U32::LE(0), |size| MessageOption::Size("pbProductId".to_string(), size.inner() as usize)),
            "pbProductId" => Vec::<u8>::new(),
            "cbLicenseInfo" => DynOption::new(U32::LE(0), |size| MessageOption::Size("pbLicenseInfo".to_string(), size.inner() as usize)),
```